REFERENCE.md

```diff
@@ -0,0 +1,746 @@
+* [`dns_mx`](#dns_mx): Retrieves DNS MX records and returns it as an array. Each record in the array will be an array of hashes with a preference and exchange field
+* [`dns_ptr`](#dns_ptr): Retrieves DNS A records and returns it as an array. Each record in the array will be a IPv4 address. An optional lambda can be given to retur
+* [`dns_srv`](#dns_srv): Retrieves DNS MX records and returns it as an array. Each record in the array will be an array of hashes with a preference and exchange field
+* [`dns_txt`](#dns_txt): Retrieves DNS TXT records and returns it as an array. Each record in the array will be a array containing the strings of the TXT record.  An 
+* [`dnsquery::a`](#dnsquery--a): Retrieves DNS A records and returns it as an array. Each record in the array will be a IPv4 address.  An optional lambda can be given to retu
+* [`dnsquery::aaaa`](#dnsquery--aaaa): Retrieves DNS AAAA records and returns it as an array. Each record in the array will be a IPv6 address.  An optional lambda can be given to r
+* [`dnsquery::cname`](#dnsquery--cname): Retrieves a DNS CNAME record and returns it as a string.  An optional lambda can be given to return a default value in case the lookup fails.
+* [`dnsquery::lookup`](#dnsquery--lookup): Do a DNS lookup and returns an array of addresses. This will follow CNAMEs and return any matching IPv4 or IPv6 addresses. See the more speci
+* [`dnsquery::mx`](#dnsquery--mx): Retrieves DNS MX records and returns it as an array. Each record in the array will be an array of hashes with a preference and exchange field
+* [`dnsquery::ptr`](#dnsquery--ptr): Retrieves DNS A records and returns it as an array. Each record in the array will be a IPv4 address. An optional lambda can be given to retur
+* [`dnsquery::rlookup`](#dnsquery--rlookup): Retrieves results from DNS reverse lookup and returns it as an array. Each record in the array will be a hostname. An optional lambda can be 
+* [`dnsquery::soa`](#dnsquery--soa): Retrieves DNS TXT records and returns it as an array. Each record in the array will be a array containing the strings of the TXT record.  An 
+* [`dnsquery::srv`](#dnsquery--srv): Retrieves DNS MX records and returns it as an array. Each record in the array will be an array of hashes with a preference and exchange field
+* [`dnsquery::txt`](#dnsquery--txt): Retrieves DNS TXT records and returns it as an array. Each record in the array will be a array containing the strings of the TXT record.  An 
+
+## Functions
+
+### <a name="dns_a"></a>`dns_a`
+
+Type: Ruby 4.x API
+
+Retrieves DNS A records and returns it as an array. Each record in the
+array will be a IPv4 address.
+
+An optional lambda can be given to return a default value in case the
+lookup fails. The lambda will only be called if the lookup failed.
+
+#### `dns_a(String $record)`
+
+The dns_a function.
+
+Returns: `Any`
+
+##### `record`
+
+Data type: `String`
+
+
+
+#### `dns_a(String $record, Callable &$block)`
```

Comment:
The callable is an interesting one. I haven't seen this before. I don't see this in https://github.com/voxpupuli/puppet-dnsquery/blob/master/lib/puppet/functions/dns_a.rb so where does it come from? is a puppet-strings bug or is this a result of the default one?

Reply:
Isn't it this? https://github.com/voxpupuli/puppet-dnsquery/blob/100c8adcf0437c16eb9a0a129c4720c1c506e24f/lib/puppet/functions/dns_a.rb#L15

Reply:
More info here. https://github.com/puppetlabs/puppet-specifications/blob/master/language/func-api.md#blocklambda-support

Reply:
Those docs suggest you could write a single `dispatch` with `optional_block_param`. That would be easier to follow.

Reply:
see #126